### DISCUSSION TITLE
Math on Coordinates

### DISCUSSION THREAD

**Discussion by fernandezc** [spectrochempy/spectrochempy#466] [Developer discussions]
Arnaud, 

For the current core refactoring, I am now reviewing the mathematical operations and functions acting on datasets and coordinates, trying to distinguish them because some functions are not useful for coordinates which, it seems to me, must be monotonic 1D functions in spectroscopies (If you see exceptions to that tell me, I will reconsider the rest). 

For example, for the coordinates, I think to keep 
the following creation functions (see https://numpy.org/doc/stable/reference/routines.array-creation.html):
- arrange
- linspace, logspace, geomspace
- fromiter 
( and possibly fromfunction)

I don't think it's useful to have also zero, ones, etc... for coordinates.
 
For mathematical operations:
- add
- subtract 
- multiply
- true_divide
- power
- reciprocal 
- log, log10, log2 
- exo, exp2 
- sqrt

do you think it is necessary to add some other useful things (for example trigonometric functions... I don't see much use)

For NDDatasets, most functions supporting mask should be implemented. 
(https://numpy.org/doc/stable/reference/routines.ma.html)

**Comment by atravert** [comment]
Hi Christian,

- For the creation: maybe `empty()`, `empty_like()` and `fromfile()` could be useful in some cases, for instance when creating a NDDataset from scratch or reading a file. 
- the energy/wavelenght dimensions are monotonic, but it is not always the case for coordinates that could be associated with 'acquisition time'  (with multiple coordinates on a given dimension): for instance a coordinate could be assoiciated with a unimodal change (e.g. heating/cooling) or even  cyclic perturbation, etc..., so specially for cyclic perturbation, maybe we could also keep the basic [trigonometric functions](https://numpy.org/doc/stable/reference/routines.math.html#trigonometric-functions) `sin()`, `cos()`, `tan()` and their inverse. Some [rounding function](https://numpy.org/doc/stable/reference/routines.math.html#)  in particular `around()` would be useful too.
- regarding masks: yes I agree, but I must confess that I don't have the reflex to use them, which is a pity because indeed this mechanism is very useful... :-)
